dev/com.ibm.ws.logging/src/com/ibm/ws/logging/internal/impl/BaseTraceService.java

```diff
@@ -1140,33 +1170,150 @@ public void publishLogRecord(LogRecord logRecord) {
-             * if messageFormat and consoleFormat have been set to "json" and "message" is a listed source.
-             * However, LogstashCollector and BluemixLogCollector will receive all messages
-             */
+            } else
+                result = LogResult.LOG;
 
-            // logSource only receives "normal" messages and messages that are not hidden.
-            if (logSource != null) {
-                publishToLogSource(routedMessage);
+            if (result == LogResult.THROTTLE_WARNING) {
+                //Do nothing to avoid duplicate throttle warnings.
+            } else if (result == LogResult.LOG) {
+                //If any messages configured to be hidden then those will not be written to console.log/message.log and redirected to trace.log.
+                if (isMessageHidden(formattedMsg)) {
+                    publishTraceLogRecord(detailLog, logRecord, NULL_ID, formattedMsg, formattedVerboseMsg);
+                    return;
+                }
+
+                /*
+                 * Messages sent through LogSource will be received by MessageLogHandler and ConsoleLogHandler
+                 * if messageFormat and consoleFormat have been set to "json" and "message" is a listed source.
+                 * However, LogstashCollector and BluemixLogCollector will receive all messages
+                 */
+
+                // logSource only receives "normal" messages and messages that are not hidden.
+                if (logSource != null) {
+                    publishToLogSource(routedMessage);
+                }
             }
         }
-
         // ODD: note that formattedMsg and formattedVerboseMsg will both be NULL if
         // this message is NOT above INFO Level.  However I believe only INFO-Level + above
         // messages are sent to this method.  So the "if (INFO-Level)" check above is probably
         // unnecessary.
 
         // Proceed to trace processing for all other log records
-        if (TraceComponent.isAnyTracingEnabled()) {
-            publishTraceLogRecord(detailLog, logRecord, NULL_ID, formattedMsg, formattedVerboseMsg);
+        if (result == LogResult.THROTTLE_WARNING) {
+            if (routedMessage != null)
+                Tr.warning(tc, "The logs are being throttled due to high volume.");
+
+        } else if (result == LogResult.LOG) {
+
+            if (TraceComponent.isAnyTracingEnabled()) {
+                publishTraceLogRecord(detailLog, logRecord, NULL_ID, formattedMsg, formattedVerboseMsg); //One trace
+            }
+
+        }
+
+    }
+
+    public enum LogResult {
+        LOG,
+        THROTTLE,
+        THROTTLE_WARNING
+    }
+
+    public LogResult logLine(Object event) {
+        LogTraceData logData = (LogTraceData) event;
+
+        String key;
+
+        if (throttleType.equals("message"))
+            key = logData.getMessage();
+        else {
+            key = logData.getMessageId();
+        }
+
+        if (key == null)
```

Comment:
Please add comments to explain why happens.

Reply:
This should never occur if a log event reaches this point but I added it in case....Already have a check later to see if key is null or not though so removed this.